File: 53.Maximum Subarray/Solution.java

public class Solution {
    public int maxSubArray(int[] nums) {
        // keep track of the minimum prefix and a running sum
        int minPreFix = 0, curSum = 0;
        int res = 0;

        for (int num : nums) {
            curSum += num;
            res = Math.max(res, curSum - minPreFix);
            minPreFix = Math.min(minPreFix, curSum);
        }

        return res;
    }
}

File: 53.Maximum Subarray/Solution.cpp
#include<iostream>
using namespace std;

int maxSubArray(vector<int>& nums) {
    // curMin tracks the minimum prefix
    // curMax tracks a running sum
    int curSum = 0, curMin = 0;
    int ret = INT_MIN;
    
    for (int num : nums) {
        curSum += num;
        ret = max(ret, curSum - curMin);
        curMin = min(curMin, curSum);
    }
    return ret;
}